internal/newusers/manager.go

```diff
@@ -0,0 +1,338 @@
+func NewManager(cacheDir string, args ...Option) (m *Manager, err error) {
+	opts := &options{
+		expirationDate:  time.Now().Add(-1 * defaultEntryExpiration),
+		cleanOnNew:      true,
+		cleanupInterval: defaultCleanupInterval,
+		procDir:         "/proc/",
+	}
+	for _, arg := range args {
+		arg(opts)
+	}
+
+	m = &Manager{
+		doClear:        make(chan struct{}),
+		quit:           make(chan struct{}),
+		cleanupStopped: make(chan struct{}),
+	}
+	c, err := cache.New(cacheDir)
+	if err != nil {
+		return nil, err
+	}
+	m.cache = c
+
+	if opts.cleanOnNew {
+		if activeUsers, err := getActiveUsers(opts.procDir); err != nil {
+			slog.Warn(fmt.Sprintf("Could not get list of active users: %v", err))
+		} else if err := m.cache.CleanExpiredUsers(activeUsers, opts.expirationDate); err != nil {
+			slog.Warn(fmt.Sprintf("Could not clean database: %v", err))
+		}
+	}
+	m.startUserCleanupRoutine(cacheDir, opts)
+
+	return m, nil
+}
+
+// Stop closes the underlying cache.
+func (m *Manager) Stop() error {
+	close(m.quit)
+	<-m.cleanupStopped
+	return m.cache.Close()
+}
+
+// UpdateUser updates the user information in the cache.
+func (m *Manager) UpdateUser(u users.UserInfo) (err error) {
+	defer decorate.OnError(&err, "failed to update user %q", u.Name)
+
+	if len(u.Groups) == 0 {
+		return fmt.Errorf("no group provided for user %s (%v)", u.Name, u.UID)
+	}
+	if u.Groups[0].GID == nil {
+		return fmt.Errorf("no gid provided for default group %q", u.Groups[0].Name)
+	}
+
+	var groupContents []cache.GroupDB
+	for _, g := range u.Groups {
+		// System group: ignore here, not part of the cache.
+		if g.GID == nil {
+			continue
+		}
+		groupContents = append(groupContents, cache.GroupDB{
+			Name: g.Name,
+			GID:  *g.GID,
+		})
+	}
+
+	// Update user information in the cache.
+	userDB := cache.UserDB{
+		UserPasswdShadow: cache.UserPasswdShadow{
+			Name:           u.Name,
+			UID:            u.UID,
+			GID:            *u.Groups[0].GID,
+			Gecos:          u.Gecos,
+			Dir:            u.Dir,
+			Shell:          u.Shell,
+			LastPwdChange:  -1,
+			MaxPwdAge:      -1,
+			PwdWarnPeriod:  -1,
+			PwdInactivity:  -1,
+			MinPwdAge:      -1,
+			ExpirationDate: -1,
+		},
+		LastLogin: time.Now(),
+	}
+	if err := m.cache.UpdateUserEntry(userDB, groupContents); err != nil {
+		return m.evaluateError(err)
+	}
+
+	return nil
+}
+
+// BrokerForUser returns the broker ID for the given user.
+func (m *Manager) BrokerForUser(username string) (string, error) {
+	brokerID, err := m.cache.BrokerForUser(username)
+	if err != nil {
+		return "", m.evaluateError(err)
+	}
+	return brokerID, err
+}
+
+// UpdateBrokerForUser updates the broker ID for the given user.
+func (m *Manager) UpdateBrokerForUser(username, brokerID string) error {
+	return m.evaluateError(m.cache.UpdateBrokerForUser(username, brokerID))
+}
+
+// UserByName returns the user information for the given user name.
+func (m *Manager) UserByName(username string) (cache.UserPasswdShadow, error) {
```

Comment:
I don’t think we should conceptually leak `cache.UserPasswdShadow` to the external API, but as you did with the rest, only a struct that is part of the new users package (or a subpackage which is common to both).

Reply:
I agree with you, we should wrap it in the "public" types.